core/node/consensus/src/storage/vm_reader.rs

```diff
@@ -0,0 +1,341 @@
+use std::time::Duration;
+
+use zksync_concurrency::ctx::Ctx;
+use zksync_contracts::load_contract;
+use zksync_node_api_server::{execution_sandbox::BlockStartInfo, tx_sender::TxSender};
+use zksync_system_constants::DEFAULT_L2_TX_GAS_PER_PUBDATA_BYTE;
+use zksync_types::{
+    api::BlockId,
+    ethabi::{Address, Contract, Function, Token},
+    fee::Fee,
+    l2::L2Tx,
+    transaction_request::CallOverrides,
+    Nonce, U256,
+};
+
+use crate::storage::ConnectionPool;
+
+pub struct VMReader {
+    pool: ConnectionPool,
+    tx_sender: TxSender,
+    consensus_authority_contract: Contract,
+    validator_registry_address: Option<Address>,
+    validator_registry_contract: Contract,
+    attester_registry_address: Option<Address>,
+    attester_registry_contract: Contract,
+}
+
+impl VMReader {
+    pub async fn new(
+        ctx: &Ctx,
+        block_id: BlockId,
+        pool: ConnectionPool,
+        tx_sender: TxSender,
+        consensus_authority_address: Address,
+    ) -> Self {
+        let consensus_authority_contract = load_contract("contracts/l2-contracts/artifacts-zk/contracts/ConsensusAuthority.sol/ConsensusAuthority.json");
+        let validator_registry_contract = load_contract("contracts/l2-contracts/artifacts-zk/contracts/ValidatorRegistry.sol/ValidatorRegistry.json");
+        let attester_registry_contract = load_contract("contracts/l2-contracts/artifacts-zk/contracts/AttesterRegistry.sol/AttesterRegistry.json");
+
+        let mut reader = Self {
+            pool,
+            tx_sender,
+            consensus_authority_contract,
+            validator_registry_address: None,
+            validator_registry_contract,
+            attester_registry_address: None,
+            attester_registry_contract,
+        };
+
+        reader.validator_registry_address = Some(
```

Comment:
It seems rather not idiomatic that the structure is created in a half-baked state and then filled.
Maybe you want to split the structure into several ones to avoid that.

Reply:
agreed - we now need to remember and rely on the fact that those `Option`s are actually `Some`s once the struct is instantiated. Additionally, it makes it harder to reason about the struct itself: I was confused why it has those addresses as `Option` - are there circumstances when they are not set?

Reply:
I agree that it is awkward, but I preferred that over having 2 structs, or not being able to utilize methods (or, transforming them into functions). I'll re-evaluate that 🙂

Reply:
Outdated / irrelevant after https://github.com/matter-labs/zksync-era/pull/2416/commits/5a1a259907b63dc7ce611e9689325f76816e56c4.